pkg/mimirtool/client/backfill.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+func (c *MimirClient) Backfill(ctx context.Context, source string, logger log.Logger) error {
+	// Scan blocks in source directory
+	es, err := os.ReadDir(source)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read directory %q", source)
+	}
+	for _, e := range es {
+		if err := c.backfillBlock(ctx, filepath.Join(source, e.Name()), logger); err != nil {
+			return err
```

Comment:
We will need to react on various errors differently, and keep going even if some blocks fail to upload.

We should give user a summary at the end of how many blocks uploaded correctly, how many already existed, and how many failed to upload.

Final exit code should depend on this too. If all blocks uploaded correctly or already existed, we can report exit code 0 (all good!). If there were some non-recoverable upload errors, let's use exit code 1.

Reply:
@pstibrany do you know how to make the Kingpin framework exit with a certain code (e.g. 1)? Should I just implement this by returning an error if one or more blocks failed to upload?

Reply:
I don't know. I would try the same (returning error) and if that fails, check Kingpin framwork documentation :)